### ISSUE TITLE
`persist_mode = false` doesn't work on window switch

### ISSUE THREAD

**Issue by DPDmancul** [akinsho/toggleterm.nvim#282]
I recall that months ago when switching between a buffer and the terminal (using the mouse or <C-w><C-w>, with <C-w> mapped to <C-\><C-n><C-w>) I found the terminal always in terminal (insert) mode.
Now when I switch window I found the terminal always in normal mode. I guessed this change was introduced by #245, so i put in my config `persist_mode = false`. It works when I close the terminal and then reopen it, but not when switching window: in that case it always returns to normal mode.

I have an alternative proposal to solve both this and #76 (which #245 aimed to solve): when pressing <C-\><C-n> and then changing window, when returning to terminal use normal mode (as #245 correctly does), but add a way to make <C-w><C-w> change window from terminal mode without pass through normal mode: in this way when returning to the terminal the terminal (insert) mode is correctly preserved by #245.  
For example, if toggle term supported insert-normal submode (C-o), the following mapping would solve:
```lua
function _G.set_terminal_keymaps ()
  for from,to in pairs {
    ["<esc>"] = [[<C-\><C-n>]],
    ["<C-w>"] = [[<C-\><C-o><C-w>]],
  } do
    vim.api.nvim_buf_set_keymap(0, "t", from, to, {
      noremap = true,
      silent = true
    })
  end
end
```
terminal mode → `<esc>` `<C-w><C-w>` ... do something ... `<C-w><C-w>` → normal mode
terminal mode → `<C-w><C-w>` ... do something ... `<C-w><C-w>` → terminal mode

**Comment by DPDmancul** [comment]
I tried to temporary solve with:
```lua
function _G.set_terminal_keymaps ()
  for from,to in pairs {
    ["<esc>"] = [[<C-\><C-n>]],
    ["<C-w><C-w>"] = [[<cmd>wincmd w<cr>]],
    ["<C-w>w"] = [[<cmd>wincmd w<cr>]],
    ["<C-w>h"] = [[<cmd>wincmd h<cr>]],
    ["<C-w>j"] = [[<cmd>wincmd j<cr>]],
    ["<C-w>k"] = [[<cmd>wincmd k<cr>]],
    ["<C-w>l"] = [[<cmd>wincmd l<cr>]],
  } do
    vim.api.nvim_buf_set_keymap(0, "t", from, to, {
      noremap = true,
      silent = true
    })
  end
end
```

I tested with both `persist_mode` to true and to false: the behavior is the same: it always returns to normal mode

**Comment by akinsho** [comment]
@DPDmancul not sure if I understand your suggested solution, but this is a known gap in the implementation I did for persist mode. It only persists the mode when toggling the buffer not on `WinLeave` and `WinEnter`. With persist mode false and no `start_in_insert` the plugin just uses vim's native behaviour which just uses normal mode for all window switching since you can't run commands/maps like that outside of normal mode.

The fix is to add the mode persistence on `WinLeave` and restore it on `WinEnter`

**Comment by akinsho** [comment]
Ok so turns out I was wrong, and it was a mixture of what I said (a bit) and another issue. One issue was that the win leave and win enter were there, but I was relying on `BufWin{Leave|Enter}` which don't actually do exactly what you'd expect.

The other issue was the recommended mappings of `<C-\><C-N>key` actually put you first in normal mode then execute the key, which means normal mode is always being persisted.

The fix I've pushed now is to use `BufEnter` to restore the mode, and I've updated the recommended mappings to what you have above i.e. `<Cmd>wincmd key` this works for me now.
